### PR TITLE
[SPARK-28252][SQL] local/global temp views should not accept duplicate column names

### PR DESCRIPTION
**Pull request by wangyum**
## What changes were proposed in this pull request?

We disabled create table with duplicate column names:
https://github.com/apache/spark/blob/f9837d3bf6c15600f926917a253bf1bdddb4d1b4/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala#L247-L250

But we can create local/global temporary views with duplicate column names. This pr disabled create local/global temporary views with duplicate column names. Otherwise:
```shell
scala> spark.sql("create temp view v1 as select 1 as col1, 2 as col1")
res0: org.apache.spark.sql.DataFrame = []

scala> spark.sql("select col1 from v1").show
19/07/04 22:27:19 WARN ObjectStore: Failed to get database global_temp, returning NoSuchObjectException
org.apache.spark.sql.AnalysisException: Reference 'col1' is ambiguous, could be: v1.col1, v1.col1.; line 1 pos 7
  at org.apache.spark.sql.catalyst.expressions.package$AttributeSeq.resolve(package.scala:259)
```

## How was this patch tested?

unit tests

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/execution/GlobalTempViewSuite.scala [hunk 1]
```diff
@@ -170,4 +170,14 @@ class GlobalTempViewSuite extends QueryTest with SharedSQLContext {
...
+
+  test("Create global temp view with duplicate column name") {
+    withGlobalTempView("src") {
+      val e = intercept[AnalysisException]{
+        sql("CREATE GLOBAL TEMP VIEW src AS SELECT 1 as a, 2 as a")
+      }
+      assert(e.getMessage.contains("Found duplicate column(s)"))
+      sql("CREATE GLOBAL TEMP VIEW src AS SELECT 1 as a, 2 as b")
+    }
+  }
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/GlobalTempViewSuite.scala]: Could you update `org.apache.spark.sql.hive.ErrorPositionSuite`, too?

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
For this case, we need a migration guide and a flag at 3.0.0 because the following has been working accidentally. Previously, users may have the following cases without noticing the duplications because they are not using those duplicated columns. (cc @gatorsmile)
```
scala> spark.version
res2: String = 2.4.3
scala> Seq((1, 1, 1)).toDF("a", "a", "b").createOrReplaceTempView("dupAttributes")
scala> sql("select b from dupAttributes").show
+---+
|  b|
+---+
|  1|
+---+
```

**Comment by wangyum** [comment]
PostgreSQL also does not accept duplicate column names:
```sql
postgres=# CREATE TEMPORARY VIEW spark_28252 as select 1 as c1, 2 as c1;
ERROR:  column "c1" specified more than once
```

File: sql/core/src/test/scala/org/apache/spark/sql/execution/GlobalTempViewSuite.scala (hunk 1 above)
Comment (wangyum) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/GlobalTempViewSuite.scala]: Fixed.

**Comment by dongjoon-hyun** [comment]
I know, @wangyum. Spark has bee worked a little differently because Spark Dataframe allows that.
> PostgreSQL also does not accept duplicate column names:

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/ErrorPositionSuite.scala [hunk 2]
```diff
@@ -34,27 +34,16 @@ class ErrorPositionSuite extends QueryTest with TestHiveSingleton with BeforeAnd
       spark.catalog.dropTempView("src")
     }
     Seq((1, "")).toDF("key", "value").createOrReplaceTempView("src")
-    Seq((1, 1, 1)).toDF("a", "a", "b").createOrReplaceTempView("dupAttributes")
```
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/ErrorPositionSuite.scala]: Hi, @gatorsmile .
Could you give us some advice for this?

**Comment by maropu** [comment]
fyi: the previous ticket for the same issue; https://issues.apache.org/jira/browse/SPARK-21363

**Comment by dongjoon-hyun** [comment]
Oh, thank you for the pointer, @maropu !

**Comment by dongjoon-hyun** [comment]
I'll add a PR link here, too.
- https://github.com/apache/spark/pull/18938 (https://issues.apache.org/jira/browse/SPARK-21363)

**Comment by dongjoon-hyun** [comment]
Hi, @gatorsmile and @rxin .
This is discussed two years ago and the decision is postponed to Apache Spark 3.0.0.
It seems to be the time to decision.

**Pull request closed without merging**